### PR TITLE
[MRG+1] Butterfly mode to raw.plot.

### PR DESCRIPTION
**Pull request by jaeilepp**
Butterfly mode now works when ``order`` is equal to ``position`` or ``selection``.

### CODE REVIEW

**Comment by agramfort** [comment]
@jaeilepp any suggestion to have this feature available without any particuliar option. ie just do raw.plot()?

**Comment by jaeilepp** [comment]
Maybe ``order='position'`` by default?

**Comment by agramfort** [comment]
without changing the API? just with key strokes?

**Comment by Eric89GXL** [comment]
> this feature available without any particuliar option. ie just do raw.plot()?

Why, are you worried about parameter bloat? The problem is that this makes scripting more difficult. It's nice to have the API option / not require interactivity. This to me is worth the added parameter. So I think we should have both a `kwarg` and keypress support.

**Comment by Eric89GXL** [comment]
Is this for #3926? If so, we could also add the `cov` option to whiten channels. Do you plan to tackle that here?

**Comment by agramfort** [comment]
you already have interactivity with raw.plot to move up and down.

people will just do raw.plot and should not have to restart the plot to see
this convenient view with butterflies

let's discuss IRL

**Comment by Eric89GXL** [comment]
@agramfort says it's okay to have `raw.plot(mode='butterfly')` be equivalent to `raw.plot()` followed by a keypress of "b". Is it doable in the code?

**Comment by jaeilepp** [comment]
Yeah, it's doable. It's just unclear what the plot should show in butterfly mode when not using channel groupings. Each channel type separately?

**Comment by agramfort** [comment]
it should toggle on the order=selection mode

**Comment by jaeilepp** [comment]
This is ready for review.

File: doc/whats_new.rst [hunk 1]
```diff
@@ -28,6 +28,8 @@ Changelog
 
     - Add new :mod:`mne.datasets.mtrf` dataset by `Chris Holdgraf`_
 
+    - Add butterfly mode (toggled with 'b' key) to :meth:`mne.io.Raw.plot by ``Jaakko Leppakangas`_
```
Comment (Eric89GXL) [inline doc/whats_new.rst]: missing backtick on `:meth:`

**Comment by jaeilepp** [comment]
> What about the kwarg we discussed above?

How about ``order='butterfly'``? Do we have any other modes?

**Comment by Eric89GXL** [comment]
Not right now, but eventually we might want `gfp` or something else. I guess `order='butterfly'` could be okay, since it sets the channel order to `selection` under the hood anyway.

@agramfort WDYT?

**Comment by Eric89GXL** [comment]
... and then please also modify an example so we can see and advertise the cool mode

**Comment by jaeilepp** [comment]
Tutorial updated: https://1512-35097627-gh.circle-artifacts.com/0/home/ubuntu/mne-python/doc/_build/html/auto_tutorials/plot_visualize_raw.html
This is ready for review.

**Comment by Eric89GXL** [comment]
LGTM +1 for merge.

I was going to say maybe the selection window shouldn't be shown in butterfly mode, but I assume it allows the user to click selections to see which sensors are highlighted on the head, which is nice even if it doesn't change the actual data plot.

**Comment by jaeilepp** [comment]
Actually clicking a channel group switches it back to 'regular' view. Basically the same as clicking 'b'.

File: mne/viz/utils.py [hunk 2]
```diff
@@ -1929,6 +1948,59 @@ def _annotation_radio_clicked(label, radio, selector):
     selector.rectprops.update(dict(facecolor=color))
 
 
+def _setup_butterfly(params):
+    """Set butterfly view of raw plotter."""
+    if 'ica' in params:
+        return
```
Comment (dengemann) [inline mne/viz/utils.py]: good catch!

File: mne/viz/utils.py [hunk 3]
```diff
@@ -1929,6 +1948,59 @@ def _annotation_radio_clicked(label, radio, selector):
...
+    butterfly = not params['butterfly']
+    ax = params['ax']
+
+    if butterfly:
+        from ..selection import _SELECTIONS
+        from .raw import _setup_browser_selection
```
Comment (dengemann) [inline mne/viz/utils.py]: is this fixing circular imports? Why is this conditional importing necessary?

**Comment by dengemann** [comment]
I am not convinced that superimposing channel types is the best idea. What can we learn from such a plot? Why don't we group it by channel types instead of "regions"?

Btw. on my computer it looks like this  by default:

screenshot 2017-04-05 23 12 02

Channel names get cut.

File: mne/viz/utils.py (hunk 3 above)
Comment (jaeilepp) [inline mne/viz/utils.py]: Yeah, it's for dealing with circular imports. I can move it at the start of the function too.

**Comment by jaeilepp** [comment]
I would say this is good for getting an overall feel of the data (spotting spurious segments etc.). Remember that you can also do picking before plotting if you only want certain channel types.

**Comment by jaeilepp** [comment]
I rotated the labels 45 degrees. I'm not sure if it's enough. Also, now clicking the selections does not switch the view. It only shows the selected region.

**Comment by Eric89GXL** [comment]
Perhaps we should set the `verticalalignment='top'` on the rotated ticks so that the end of each selection name aligns with that selection:

https://1521-35097627-gh.circle-artifacts.com/0/home/ubuntu/mne-python/doc/_build/html/auto_tutorials/plot_visualize_raw.html

**Comment by jaeilepp** [comment]
Codecov failure seems unrelated.

**Comment by dengemann** [comment]
>I would say this is good for getting an overall feel of the data (spotting
spurious segments etc.). >Remember that you can also do picking before
plotting if you only want certain channel types.

@jailepp maybe a well chosen transparency value could enhance the
accessibility of this new plotting option. To be tried.

**Comment by jaeilepp** [comment]
You mean making some channel types having <1 alpha? Yeah, I can experiment.

**Comment by jaeilepp** [comment]
Looks like changing the alpha value makes little difference, but perhaps the z-value change makes more sense. I'm wondering if it's usually so that grad amplitudes (scaling applied) are bigger than mag amplitudes, so it makes sense plotting mag in front of grad.
alpha

**Comment by dengemann** [comment]
Yes the grads are denoising in nature. Looks already much better!

**Comment by Eric89GXL** [comment]
Still +1 for merge from me. @dengemann you have a -1 are you happy or at least okay with it now?

https://1535-35097627-gh.circle-artifacts.com/0/home/ubuntu/mne-python/doc/_build/html/auto_tutorials/plot_visualize_raw.html#drawing-annotations

**Comment by dengemann** [comment]
I am way more happy now. But you need a rebase @jaeilepp.

**Comment by Eric89GXL** [comment]
I resolved the conflicts with the GUI, feel free to merge if you're happy @dengemann

**Comment by dengemann** [comment]
let me pull and play. Then I give you more feedback or merge :)

**Comment by dengemann** [comment]
@jaeilepp @Eric89GXL the sensor selection view is only for displaying the selection or does it have any effect on the data view? If it is only for viewing the selection it could be visibly stated somewhere. And what are the two red sensors which are visible in any of the selections?

**Comment by dengemann** [comment]
Other then then the view problem has been fixed, thanks @jaeilepp @Eric89GXL

**Comment by Eric89GXL** [comment]
>  the sensor selection view is only for displaying the selection or does it have any effect on the data view?

See discussion above, no effect on the view.

We could make it change the color of the selected name in the plot. Then clicking would cause a color change in the plot, making it clear that it doesn't change anything else in the plot.

**Comment by dengemann** [comment]
> We could make it change the color of the selected name in the plot. Then clicking would cause a color change in the plot, making it clear that it doesn't change anything else in the plot.

or at least change the font of the selected group as a small pointer. Then it would be clear I think.

**Comment by Eric89GXL** [comment]
You mean the type face? I think that would look weird. Text color is something we change more often anyway.

**Comment by dengemann** [comment]
Well whatever works. Or underline or highlight in some way. Just have
something subtly but visibly changing.

**Comment by jaeilepp** [comment]
> And what are the two red sensors which are visible in any of the selections?

Red circle edges on the selection window indicate a bad channel.

**Comment by dengemann** [comment]
Ahh good. Make sure this is commented on in the docs.

**Comment by jaeilepp** [comment]
> We could make it change the color of the selected name in the plot. Then clicking would cause a color change in the plot, making it clear that it doesn't change anything else in the plot.

AFAIK this would require redraw of the whole plot, making it laggy. Not worth it IMO.

**Comment by Eric89GXL** [comment]
Agreed

**Comment by jaeilepp** [comment]
Had a discussion with @agramfort . Plan is to separate the order param to ``picks`` and ``group_by`` to enable different kinds of groupings (position, channel type).

File: mne/viz/raw.py [hunk 4]
```diff
@@ -178,6 +173,17 @@ def plot_raw(raw, events=None, duration=10.0, start=0.0, n_channels=20,
...
+        How to group channels. 'type' groups by channel type, 'original'
+        plots in the order of ch_names, 'selection' uses Elekta's channel
+        groupings (only works for Neuromag data), 'position' groups the
+        channels by the positions of the sensors. 'selection' and 'position'
+        modes allow butterfly mode (press 'b') and custom selections by using
+        lasso selector on the topomap. Pressing ``ctrl`` key while selecting
+        allows appending to the current selection. Channels marked as bad
+        appear with red edges on the topomap. 'butterfly' is equal to
+        'position', but the plotter is started in butterfly mode. 'type' and
+        'original' groups the channels by type in butterfly mode.
```
Comment (jaeilepp) [inline mne/viz/raw.py]: Can't remember if this is exactly as we planned, but see if it's okay.

Comment (Eric89GXL) [inline mne/viz/raw.py]: There is no way to start `butterfly` mode with `'position'`.

It seems like we're trying to force a `butterfly` param into the `group_by` when it should really be separate.

Both `order` and `group_by` affect channel order currently. Why not leave `order` allowing string_types and thus the only thing that sets channel order, and then also allow `butterfly=True | False`? Seems cleaner to me.

Comment (jaeilepp) [inline mne/viz/raw.py]: One benefit of having it like this is the possibility to limit the amount of channels plotted (for instance ``group_by='position', order=meg_picks``). This is useful with large amount of channels (or with noisy channels) as plotting everything in butterfly mode can become quite laggy.

Comment (Eric89GXL) [inline mne/viz/raw.py]: So what we really probably need is `picks` (list of int), `order` (string only), and `butterfly` (bool). It's more parameters, but trying to mix them is leading to a bad API.

Comment (jaeilepp) [inline mne/viz/raw.py]: Agreed.

**Comment by jaeilepp** [comment]
I think I've fixed all the bugs, but there are so many combinations with these new parameters that it's possible there are still some left. Give it a try!

File: tutorials/plot_visualize_raw.py [hunk 5]
```diff
@@ -70,11 +70,12 @@
...
-# groups the channels the same way as the position mode. To toggle between
-# these modes, press 'b' key when the plotter window is active. Notice that
-# ``group_by`` also effects the channel groupings in butterfly mode.
-raw.plot(group_by='butterfly')
+# to customize the order and select a subset of channels for plotting (picks).
+# Here we use the butterfly mode and group the channels by position. To toggle
+# between regular and butterfly modes, press 'b' key when the plotter window is
+# active. Notice that ``group_by`` also affects the channel groupings in
+# butterfly mode.
+raw.plot(butterfly=True)
```
Comment (mmagnuski) [inline tutorials/plot_visualize_raw.py]: according to the text above there should be a `group_by='position'` here, right?

Comment (jaeilepp) [inline tutorials/plot_visualize_raw.py]: True.

**Comment by agramfort** [comment]
In butterfly mode I see an ECG block for sample although this is no ECG in the sample data.

help window in not updated.

**Comment by agramfort** [comment]
Playing with the raw.plot I also see in my console:

IndexError: index 21 is out of bounds for axis 1 with size 21

obtained when running the viz tutorial for raw object.

**Comment by agramfort** [comment]
when clicking the green line should appear on top of the lines.

**Comment by agramfort** [comment]
@jaeilepp @Eric89GXL do you think that we could subsample automatically based on time interval visualized? I think this could potentially greatly accelerate the UI which is lagging easily for a real MEG dataset.

**Comment by Eric89GXL** [comment]
Not without potentially missing artifacts or spikey signals, so I think
it's unsafe

**Comment by agramfort** [comment]
do you see a way to make the viz more responsive?

**Comment by mmagnuski** [comment]
My experience with lags in the gui is that for raw it works much more smoothly than for epoched. I one had a raw recording segmented into thousands of 1s epochs - the lag in the gui response was measured in seconds (about 5 or 6)  which made it practically useless. What was strange - the lag was irrespective of how many epochs were displayed, only the total number seemed to affect the lag. If this was an accurate observation - probably there is substantial improvement for epochs plot possible.

**Comment by agramfort** [comment]
can we profile this using a script by faking interactions?

**Comment by Eric89GXL** [comment]
> do you see a way to make the viz more responsive?

Not for `raw` unfortunately, other than changing rendering libraries at some point.

> can we profile this using a script by faking interactions?

Hopefully it is just `fig.canvas.key_press_event('right')`

**Comment by jaeilepp** [comment]
> Playing with the raw.plot I also see in my console:
IndexError: index 21 is out of bounds for axis 1 with size 21
obtained when running the viz tutorial for raw object.

Unable to reproduce. Any idea how you got it?

> help window in not updated.

Are you sure you have the latest changes? It should be there.

**Comment by jaeilepp** [comment]
> What was strange - the lag was irrespective of how many epochs were displayed, only the total number seemed to affect the lag

I noticed that ``__getitem__`` of epochs that is utilized when fetching a subset for plotting is far from optimal (making a copy of the object). I'll look into it.

**Comment by Eric89GXL** [comment]
You have conflicts, can you rebase?

**Comment by jaeilepp** [comment]
Rebased.

File: mne/viz/raw.py [hunk 6]
```diff
@@ -183,6 +179,17 @@ def plot_raw(raw, events=None, duration=10.0, start=0.0, n_channels=20,
...
         to modify the data stored in the Raw object.
+    group_by : str
+        How to group channels. 'type' groups by channel type, 'original'
+        plots in the order of ch_names, 'selection' uses Elekta's channel
+        groupings (only works for Neuromag data), 'position' groups the
+        channels by the positions of the sensors. 'selection' and 'position'
+        modes allow custom selections by using lasso selector on the topomap.
+        Pressing ``ctrl`` key while selecting allows appending to the current
+        selection. Channels marked as bad appear with red edges on the topomap.
+        'type' and 'original' groups the channels by type in butterfly mode.
```
Comment (Eric89GXL) [inline mne/viz/raw.py]: There is no mention of how this interacts with `order`. Which takes precedence if both are given?

You should probably mention `group_by` in the `order` description and vice-versa.

**Comment by agramfort** [comment]
some UI bugs:

- when I press Esc when reading the help window the window does not disappear. Esc on the main window closes the window but the help window stays open.

- when I toggle OFF the avg proj of EEG. Then close the proj window and show it again. The EEG avg proj appears ON although it's not OFF.

- when I do: raw.plot(order=np.arange(306)) and toggle on the butterfly I see eeg, eog and MISC although I should only see MEG. It's the same if I do `raw.plot(order=np.arange(306), group_by='selection')` in butterfly mode, it shows me the EEG selections :( 

@jaeilepp let me know when I need to try again.

thx

**Comment by jaeilepp** [comment]
I believe I have addressed the comments.

**Comment by agramfort** [comment]
@jaeilepp I still see misc channels when I do:

raw.plot(order=np.arange(306), group_by='selection')

besides works great for me. Any other beta tester?

**Comment by Eric89GXL** [comment]
I can look tomorrow if nobody else beats me to it

**Comment by Eric89GXL** [comment]
Got sidetracked, will have to look tomorrow

**Comment by Eric89GXL** [comment]
I like the aesthetics, functionality, and API.

Scrolling in time in butterfly mode feels at least twice as slow as in non-butterfly mode. I guess there is no getting around this because it's also that slow when doing e.g. `raw.plot(n_channels=306)`, so it's just due to having so many lines to plot.

I wonder if having some alpha like 0.5 on the lines in butterfly mode would make the plots more informative (don't we do this with `plot_psd` with average=False mode?), but perhaps you tried that already.

**Comment by jaeilepp** [comment]
Yeah, I played with different alpha values and it made very little difference to the visuals. I also tried optimizing the plotting by trying to only update the necessary parts of the plot, but failed to make any speedups. I guess it's already optimized at the ``draw`` function of mpl.

**Comment by Eric89GXL** [comment]
+1 for merge from me

**Comment by agramfort** [comment]
thanks @jaeilepp !

**Pull request merged**